query-parser/src/parse.rs

```diff
@@ -10,156 +10,191 @@
-                let r: ParseResult<FindSpec, _> = p.parse_lazy(&y[..]).into();
-                FindSp::to_parsed_value(r)
-            })
-            .parse_stream(input)
-    }
-
-    fn elements() -> FindSpParser<Vec<Element>, I> {
-        fn_parser(FindSp::<I>::elements_, "elements")
-    }
-
-    fn elements_(input: I) -> ParseResult<Vec<Element>, I> {
-        (many1::<Vec<Variable>, _>(FindSp::variable()), eof())
-            .map(|(vars, _)| {
-                vars.into_iter()
-                    .map(Element::Variable)
-                    .collect()
-            })
-            .parse_stream(input)
-    }
+#[derive(Clone,Debug,Eq,PartialEq)]
+pub enum QueryParseError {
+    InvalidInput(edn::Value),
+    EdnParseError(edn::parse::ParseError),
+    MissingField(edn::Keyword),
+    FindParseError(FindParseError),
+}
 
-    fn find_rel() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_rel_, "find_rel")
+impl From<edn::parse::ParseError> for QueryParseError {
+    fn from(err: edn::parse::ParseError) -> QueryParseError {
+        QueryParseError::EdnParseError(err)
     }
+}
 
-    fn find_rel_(input: I) -> ParseResult<FindSpec, I> {
-        FindSp::elements().map(FindSpec::FindRel).parse_stream(input)
-    }
+pub type WhereParseResult = Result<Vec<WhereClause>, WhereParseError>;
+pub type FindParseResult = Result<FindSpec, FindParseError>;
+pub type QueryParseResult = Result<FindQuery, QueryParseError>;
 
-    fn find_tuple() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_tuple_, "find_tuple")
-    }
 
-    fn find_tuple_(input: I) -> ParseResult<FindSpec, I> {
-        satisfy_unwrap!(edn::Value::Vector, y, {
-                let r: ParseResult<FindSpec, _> =
-                    FindSp::elements().map(FindSpec::FindTuple).parse_lazy(&y[..]).into();
-                FindSp::to_parsed_value(r)
-            })
-            .parse_stream(input)
-    }
-
-    fn find() -> FindSpParser<FindSpec, I> {
-        fn_parser(FindSp::<I>::find_, "find")
-    }
-
-    fn find_(input: I) -> ParseResult<FindSpec, I> {
-        // Any one of the four specs might apply, so we combine them with `choice`.
-        // Our parsers consume input, so we need to wrap them in `try` so that they
-        // operate independently.
-        choice::<[&mut Parser<Input = I, Output = FindSpec>; 4],
-                 _>([&mut try(FindSp::find_scalar()),
-                     &mut try(FindSp::find_coll()),
-                     &mut try(FindSp::find_tuple()),
-                     &mut try(FindSp::find_rel())])
-            .parse_stream(input)
-    }
+pub struct Query<I>(::std::marker::PhantomData<fn(I) -> I>);
 
+impl<I> Query<I>
+    where I: Stream<Item = edn::Value>
+{
     fn to_parsed_value<T>(r: ParseResult<T, I>) -> Option<T> {
         r.ok().map(|x| x.0)
     }
 }
 
+def_value_satisfy_parser_fn!(Query, variable, Variable, Variable::from_value);
```

Comment:
Yeah, it's pretty verbose to get to where parsing code in most other languages ends up, which is:

```
(or (parse-var in)
    (parse-constant in)
    (error "Expected var or constant"))
```